### PR TITLE
use bazel 8 (bzlmod migration) 

### LINKED ISSUE #131: Cannot build using bazel 8
```
❯ bazel build //:fizzbee
2024/12/13 20:52:44 Downloading https://releases.bazel.build/8.0.0/release/bazel-8.0.0-darwin-arm64...
Downloading: 56 MB out of 56 MB (100%) 
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
WARNING: Target pattern parsing failed.
ERROR: Skipping '//:fizzbee': error loading package '': Unable to find package for @@[unknown repo 'io_bazel_rules_go' requested from @@]//go:def.bzl: The repository '@@[unknown repo 'io_bazel_rules_go' requested from @@]' could not be resolved: No repository visible as '@io_bazel_rules_go' from main repository. Was the repository introduced in WORKSPACE? The WORKSPACE file is disabled by default in Bazel 8 (late 2024) and will be removed in Bazel 9 (late 2025), please migrate to Bzlmod. See https://bazel.build/external/migration.
ERROR: error loading package '': Unable to find package for @@[unknown repo 'io_bazel_rules_go' requested from @@]//go:def.bzl: The repository '@@[unknown repo 'io_bazel_rules_go' requested from @@]' could not be resolved: No repository visible as '@io_bazel_rules_go' from main repository. Was the repository introduced in WORKSPACE? The WORKSPACE file is disabled by default in Bazel 8 (late 2024) and will be removed in Bazel 9 (late 2025), please migrate to Bzlmod. See https://bazel.build/external/migration.
INFO: Elapsed time: 16.963s
INFO: 0 processes.
ERROR: Build did NOT complete successfully
```

### PR DESCRIPTION
**Pull request by tekumara**
Migrates to bzlmod which is required by Bazel 8 see https://bazel.build/external/migration

Go deps are now defined in go.mod

Resolves #131

### CODE REVIEW

File: MODULE.bazel [hunk 1]
```diff
@@ -1,6 +1,46 @@
...
+    "net_starlark_go",
+    "org_golang_google_protobuf",
+    "org_golang_x_sys",
+)
+
+# Proto
+bazel_dep(name = "rules_proto", version = "5.3.0-21.7")
+bazel_dep(
+    name = "protobuf",
+    version = "26.0",
```
Comment (tekumara) [inline MODULE.bazel:30]: previously we were using 25.2 but that doesn't exist on [bazel central](https://registry.bazel.build/modules/protobuf)

**Comment by tekumara** [comment]
NB: This PR has lots of WIP commits - I recommend squash merging it.

File: modelchecker/BUILD.bazel [hunk 2]
```diff
@@ -22,15 +22,13 @@ go_library(
...
     deps = [
         "//lib",
-        "//proto",
-        "//proto:options",
+        "//proto:proto_go_proto",
         "@com_github_golang_glog//:glog",
-        "@com_github_jayaprabhakar_go_clone//:go-clone",
-        "@org_golang_google_protobuf//encoding/protojson:go_default_library",
-        "@org_golang_google_protobuf//proto:go_default_library",
+        "@com_github_huandu_go_clone//:go-clone",
```
Comment (jp-fizzbee) [inline modelchecker/BUILD.bazel:27]: @tekumara One reason I had to fork the cloning library was it was missing a feature to override pointer references. 
https://github.com/huandu/go-clone/issues/23

**Comment by jayaprabhakar** [comment]
Thanks for the PR. I only have one issue. Is there a reason to switch one of the dependency?

File: modelchecker/BUILD.bazel (hunk 2 above)
Comment (tekumara) [reply modelchecker/BUILD.bazel:27]: It's still using your fork - see the replace statement in go.mod 

I had to do it like this to make go.mod work

Comment (tekumara) [reply modelchecker/BUILD.bazel:27]: https://go.dev/ref/mod#go-mod-file-replace

Comment (tekumara) [reply modelchecker/BUILD.bazel:27]: I'm assuming bazel honours this but admittedly I haven't tested this - is there a way to test this?

**Review by jp-fizzbee — APPROVED** [review]

**Comment by jp-fizzbee** [comment]
I ran it, I see this error.

```
Starting local Bazel server and connecting to it...
WARNING: For repository 'protobuf', the root module requires module version protobuf@26.0, but got protobuf@29.0 in the resolved dependency graph. Please update the version in your MODULE.bazel or set --check_direct_dependencies=off
WARNING: Couldn't auto load rules or symbols, because no dependency on module/repository 'rules_android' found. This will result in a failure if there's a reference to those rules or symbols.
INFO: Analyzed 2 targets (135 packages loaded, 5344 targets configured).
INFO: From Linking external/protobuf+/src/google/protobuf/io/libio_win32.a [for tool]:
warning: /Library/Developer/CommandLineTools/usr/bin/libtool: archive library: bazel-out/darwin_arm64-opt-exec-ST-d57f47055a04/bin/external/protobuf+/src/google/protobuf/io/libio_win32.a the table of contents is empty (no object file members in the library define global symbols)
ERROR: /private/var/tmp/_bazel_jp/44f378705132ae1485a21c4be97e06a5/external/protobuf+/src/google/protobuf/io/BUILD.bazel:99:11: Compiling src/google/protobuf/io/printer.cc [for tool] failed: (Exit 1): cc_wrapper.sh failed: error executing CppCompile command (from target @@protobuf+//src/google/protobuf/io:printer) external/rules_cc++cc_configure_extension+local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object ... (remaining 50 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from external/protobuf+/src/google/protobuf/io/printer.cc:12:
bazel-out/darwin_arm64-opt-exec-ST-d57f47055a04/bin/external/protobuf+/src/google/protobuf/io/_virtual_includes/printer/google/protobuf/io/printer.h:918:19: error: 'get<std::function<bool ()>, std::string, std::function<bool ()>>' is unavailable: introduced in macOS 10.13
    value = absl::get<Callback>(that.value);
                  ^
bazel-out/darwin_arm64-opt-exec-ST-d57f47055a04/bin/external/protobuf+/src/google/protobuf/io/_virtual_includes/printer/google/protobuf/io/printer.h:863:11: note: in instantiation of function template specialization 'google::protobuf::io::Printer::ValueImpl<false>::operator=<true>' requested here
    *this = that;
          ^
bazel-out/darwin_arm64-opt-exec-ST-d57f47055a04/bin/external/protobuf+/src/google/protobuf/io/_virtual_includes/printer/google/protobuf/io/printer.h:1150:12: note: in instantiation of function template specialization 'google::protobuf::io::Printer::ValueImpl<false>::ValueImpl<true>' requested here
    return ValueView(it->second);
           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/variant:1577:22: note: 'get<std::function<bool ()>, std::string, std::function<bool ()>>' has been explicitly marked unavailable here
constexpr const _Tp& get(const variant<_Types...>& __v) {
                     ^
1 error generated.
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 20.992s, Critical Path: 2.82s
INFO: 704 processes: 458 internal, 246 darwin-sandbox.
ERROR: Build did NOT complete successfully
⏺ jp@ /Users/jp/src1/fizzbee (main) # 


```

**Comment by jayaprabhakar** [comment]
It looks like this is an issue with Mac. 
https://github.com/protocolbuffers/protobuf/issues/16944

The current workaround should be to add 
```
build --host_cxxopt=-std=c++14 --cxxopt=-std=c++14
```
To `.bazelrc`

I'm not sure if this works in Linux.

**Comment by tekumara** [comment]
Interesting I didn't hit that on my Mac.

I've tried the PR without that .bazelrc on Ubuntu arm64 and it built fine.

**Comment by tekumara** [comment]
Interesting I didn't hit that on my Mac.

I notice we are using an old version of protobuf.

I've tried the PR without that .bazelrc on Ubuntu arm64 and it built fine.

**Comment by jp-fizzbee** [comment]
For now, I added `build --host_cxxopt=-std=c++14 --cxxopt=-std=c++14` to my local .bazelrc to make it work. I am not sure if it is an issue on mine or others too.

**Pull request merged**